mapeditor/mainwindow.cpp

```diff
@@ -0,0 +1,1115 @@
+}
+
+void MainWindow::on_actionMapSettings_triggered()
+{
+	auto settingsDialog = new MapSettings(controller, this);
+	settingsDialog->setWindowModality(Qt::WindowModal);
+	settingsDialog->setModal(true);
+}
+
+
+void MainWindow::on_actionPlayers_settings_triggered()
+{
+	auto settingsDialog = new PlayerSettings(controller, this);
+	settingsDialog->setWindowModality(Qt::WindowModal);
+	settingsDialog->setModal(true);
+	connect(settingsDialog, &QDialog::finished, this, &MainWindow::onPlayersChanged);
+}
+
+QAction * MainWindow::getActionPlayer(const PlayerColor & player)
+{
+	if(player.getNum() == 0) return ui->actionPlayer_1;
+	if(player.getNum() == 1) return ui->actionPlayer_2;
+	if(player.getNum() == 2) return ui->actionPlayer_3;
+	if(player.getNum() == 3) return ui->actionPlayer_4;
+	if(player.getNum() == 4) return ui->actionPlayer_5;
+	if(player.getNum() == 5) return ui->actionPlayer_6;
+	if(player.getNum() == 6) return ui->actionPlayer_7;
+	if(player.getNum() == 7) return ui->actionPlayer_8;
+	return ui->actionNeutral;
+}
+
+void MainWindow::switchDefaultPlayer(const PlayerColor & player)
+{
+	if(controller.defaultPlayer == player)
+		return;
+	
+	ui->actionNeutral->blockSignals(true);
+	ui->actionNeutral->setChecked(PlayerColor::NEUTRAL == player);
+	ui->actionNeutral->blockSignals(false);
+	for(int i = 0; i < 8; ++i)
+	{
+		getActionPlayer(PlayerColor(i))->blockSignals(true);
+		getActionPlayer(PlayerColor(i))->setChecked(PlayerColor(i) == player);
+		getActionPlayer(PlayerColor(i))->blockSignals(false);
+	}
+	controller.defaultPlayer = player;
+}
+
+void MainWindow::onPlayersChanged()
+{
+	if(controller.map())
+	{
+		getActionPlayer(PlayerColor::NEUTRAL)->setEnabled(true);
+		for(int i = 0; i < controller.map()->players.size(); ++i)
+			getActionPlayer(PlayerColor(i))->setEnabled(controller.map()->players.at(i).canAnyonePlay());
+		if(!getActionPlayer(controller.defaultPlayer)->isEnabled() || controller.defaultPlayer == PlayerColor::NEUTRAL)
+			switchDefaultPlayer(PlayerColor::NEUTRAL);
+	}
+	else
+	{
+		for(int i = 0; i < PlayerColor::PLAYER_LIMIT.getNum(); ++i)
+			getActionPlayer(PlayerColor(i))->setEnabled(false);
+		getActionPlayer(PlayerColor::NEUTRAL)->setEnabled(false);
+	}
+	
+}
+
+
+
+void MainWindow::enableUndo(bool enable)
+{
+	ui->actionUndo->setEnabled(enable);
+}
+
+void MainWindow::enableRedo(bool enable)
+{
+	ui->actionRedo->setEnabled(enable);
+}
+
+void MainWindow::onSelectionMade(int level, bool anythingSelected)
+{
+	if (level == mapLevel)
+	{
+		auto info = QString::asprintf("Selection on layer %d: %b", level, anythingSelected ? "true" : "false");
+		setStatusMessage(info);
+
+		ui->actionErase->setEnabled(anythingSelected);
+		ui->toolErase->setEnabled(anythingSelected);
+	}
+}
+void MainWindow::displayStatus(const QString& message, int timeout /* = 2000 */)
+{
+	statusBar()->showMessage(message, timeout);
+}
+
+void MainWindow::on_actionValidate_triggered()
+{
+	new Validator(controller.map(), this);
```

Comment:
Who will free it afterwards?

Reply:
qt objects are destroyed by themselves (on close on parent close)

Reply:
This won't get deleted because it's not assigned an owner; it'll just be dangling.

Reply:
@AgostonSzepessy , please get familiar with QT documentation

Reply:
Hey @Nordsoft91, in the future, I would appreciate it if you were less passive-aggressive, but instead more polite and helpful. This comment wasn't really either of those.

Reply:
I've already explained above that qt manages memory, you statement was just incorrect so I suggest to get details from docs (qt has very good docs actually). What else is expected? Sorry, nothing offensive anyway, I value any kind of feedback.

Reply:
I agree. There was nothing aggressive. @AgostonSzepessy you did not ask for any advanced clarification but instead insisted on @Nordsoft91 being wrong without any additional information regarding why you think so. Or is there something QT specific under **because it's not assigned an owner**? How should this be done?

Reply:
@Nordsoft91 thanks for clarifying, it makes sense now. I just found your comment very abrupt and not very helpful. However, all is good and I have no ill feelings.

@nullkiller I did provide my reasoning, but it's okay now; water under the bridge.
This is something Qt specific. Anything that derives from a QObject (Validator derives from QDialog) will get cleaned up automatically by the Qt framework when the thing they're attached to also get deleted/removed. In this case, it'll get deleted when either Validator is closed or MainWindow is closed.